NAS-127114 / None / NAS-127114 - Add support for alternate data streams

anodos325 opened the pull request:
SMB Protocol Background:
------------------------
Filesystems presented over the SMB protocol may support alternate data streams ("named streams") within a file or a directory. This support is designated by the filesystem attribute FILE_NAMED_STREAMS. Named streams are not identical to extended attributes (EAs), which may also be supported by the same SMB server.

A named stream is a place within a file in addition to the main stream (normal file data) where data is stored. Named streams have different data than the main stream (and than each other) and may be written and read independenly of each other. Named streams for a file are designated by appending a ":" colon character to the file name followed by the name of the alternate data stream. Stream names may be no more than 255 characters in length and are subject to the characteristics and limitations documented in MS-FSCC Section 2.1.5 Pathname and following.

A list of named streams for a file can be gathered by submitting an SMB2_QUERY_INFO request for FILE_STREAM_INFORMATION. The expected server response is documented in MS-FSS Section 2.4.43 FileStreamInformation.

Streams are typically smallish in size (less than 200 bytes individually), and are rarely used apart from MacOS SMB clients.

TrueNAS /ZFS background:
------------------------
Solaris supported a similar feature set through its file-backed xattr capabilities and APIs. This meant that the kernel SMB server in solaris was able to seamlessly provide support for named streams. When ZFS was ported to FreeBSD and Linux the extattr and xattr OS APIs were layered on top of the ZFS file-backed xattrs. As time progressed and ZFS on Linux saw more use, it was determined that the performance and lack of atomocity of operations on file-backed xattrs was insufficient for some application requirements (this was especially the case for Samba shares), this eventually led to the ZFS dataset configuration parameter for System-Attribute (SA) backed xattrs on Linux (which is the TrueNAS default). With this configuration, xattrs up to a certain size are written as SA, and larger xattrs are written as files. The practical result of this is that TrueNAS can support extended attributes that are much greater in size than a traditional Linux file server. Unforutnately, due to inability to perform partial reads and writes on extended attributes a 2 MiB upper bound is placed as the maximum size of a single extended attribute / named stream in TrueNAS.

Samba background:
-----------------
Samba has the ability to present extended attributes as named streams to SMB clients. This is achieved by prepending a special prefix to the exended attribute (to differentiate the streams xattrs from normal xattrs that are presented as EAs over the SMB protocol). Due to historical design decisions, the Samba module in charge of translating xattrs into streams appends an extra NULL byte to the xattr on writes to the local filesystem and strips it off when converting to a stream for SMB clients.

Implementation details:
-----------------------
This commit adds support for the Linux kernel SMB2/3 client to enumerate streams on a remote SMB server by including them in the output of listxattr with the special Samba prefix. Streams may be written to the remote SMB server via setxattr and read through getxtattr. The Samba-specific behavior for appending / removing an extra byte to the xattr can be disabled by setting /proc/fs/cifs/stream_samba_compat to 0.

anodos325 committed:
NAS-127114 - Add support for alternate data streams

SMB Protocol Background:
------------------------
Filesystems presented over the SMB protocol may support alternate data
streams ("named streams") within a file or a directory. This support is
designated by the filesystem attribute FILE_NAMED_STREAMS. Named streams
are not identical to extended attributes (EAs), which may also be
supported by the same SMB server.

A named stream is a place within a file in addition to the main stream
(normal file data) where data is stored. Named streams have different
data than the main stream (and than each other) and may be written and
read independenly of each other. Named streams for a file are designated
by appending a ":" colon character to the file name followed by the
name of the alternate data stream. Stream names may be no more than 255
characters in length and are subject to the characteristics and
limitations documented in MS-FSCC Section 2.1.5 Pathname and following.

A list of named streams for a file can be gathered by submitting an
SMB2_QUERY_INFO request for FILE_STREAM_INFORMATION. The expected server
response is documented in MS-FSS Section 2.4.43 FileStreamInformation.

Streams are typically smallish in size (less than 200 bytes individually),
and are rarely used apart from MacOS SMB clients.

TrueNAS /ZFS background:
------------------------
Solaris supported a similar feature set through its file-backed xattr
capabilities and APIs. This meant that the kernel SMB server in solaris
was able to seamlessly provide support for named streams. When ZFS was
ported to FreeBSD and Linux the extattr and xattr OS APIs were layered
on top of the ZFS file-backed xattrs. As time progressed and ZFS on
Linux saw more use, it was determined that the performance and lack of
atomocity of operations on file-backed xattrs was insufficient for
some application requirements (this was especially the case for Samba
shares), this eventually led to the ZFS dataset configuration parameter
for SA-backed xattrs on Linux (which is the TrueNAS default). With this
configuration, xattrs up to a certain size are written as SA, and larger
xattrs are written as files. The practical result of this is that
TrueNAS can support extended attributes that are much greater in size
than a traditional Linux file server. Unforutnately, due to inability
to perform partial reads and writes on extended attributes a 2 MiB
upper bound is placed as the maximum size of a single extended attribute
/ named stream in TrueNAS.

Samba background:
-----------------
Samba has the ability to present extended attributes as named streams
to SMB clients. This is achieved by prepending a special prefix to
the exended attribute (to differentiate the streams xattrs from normal
xattrs that are presented as EAs over the SMB protocol). Due to
historical design decisions, the Samba module in charge of translating
xattrs into streams appends an extra NULL byte to the xattr on writes
to the local filesystem and strips it off when converting to a stream
for SMB clients.

Implementation details:
-----------------------
This commit adds support for the Linux kernel SMB2/3 client to enumerate
streams on a remote SMB server by including them in the output of
listxattr with the special Samba prefix. Streams may be written to
the remote SMB server via setxattr and read through getxtattr. The
Samba-specific behavior for appending / removing an extra byte to
the xattr can be disabled by setting /proc/fs/cifs/stream_samba_compat
to 0.

bugclerk commented:
Jira URL: https://ixsystems.atlassian.net/browse/NAS-127114

mgrimesix commented on fs/smb/client/smb2misc.c:
```diff
@@ -482,6 +482,84 @@ cifs_convert_path_to_utf16(const char *from, struct cifs_sb_info *cifs_sb)
+get_start_of_path(const char *from, struct cifs_sb_info *cifs_sb)
+{
+	/* Windows doesn't allow paths beginning with \ */
+	if (from[0] == '\\')
```
Maybe make sure all '\\' are removed?
```
char* ptr = from;
while (*ptr++ == '\\')
if (ptr != from)
    return ptr;
```
Snippet not checked for correctness.
Maybe similar for the leading slash below?

anodos325 replied:
The remainder of string gets normalized in the function that converts to UTF16. The new code basically duplicates the string conversion function above it (but for named streams). We have to have new function because the colon is a valid character (separator) for the stream and so we have to convert the filename and stream name separately. I'll add a comment to this effect and maybe simplify here.

mgrimesix commented on fs/smb/client/smb2misc.c:
```diff
@@ -482,6 +482,84 @@ cifs_convert_path_to_utf16(const char *from, struct cifs_sb_info *cifs_sb)
+
+	memcpy(buf, file_buf, file_len);
+
+	// cifs_strndup_to_utff16 appends two zero bytes
```
Just to be certain... 'utff16' is a typo?

anodos325 replied:
Yes

anodos325 committed:
Expand comments

bmeagherix commented on fs/smb/client/xattr.c:
```diff
@@ -159,7 +160,13 @@ static int cifs_xattr_set(const struct xattr_handler *handler,
+		    (strncmp(name, STREAM_XATTR, STREAM_XATTR_LEN) == 0)) {
+			rc = set_stream_xattr(dentry, full_path, xid, pTcon,
+					      name, value, size, flags);
+#endif /* CONFIG_TRUENAS */
+		} else if (pTcon->ses->server->ops->set_EA) {
```
Maybe move the `} else` portion inside the `#ifdef ... #endif` block.  Something like

```
#ifdef CONFIG_TRUENAS
		if ((pTcon->ses->server->dialect > SMB30_PROT_ID) &&
		    (strncmp(name, STREAM_XATTR, STREAM_XATTR_LEN) == 0)) {
			rc = set_stream_xattr(dentry, full_path, xid, pTcon,
					      name, value, size, flags);
		} else
#endif /* CONFIG_TRUENAS */
		if (pTcon->ses->server->ops->set_EA) {
```

anodos325 replied:
Clever

bmeagherix commented on fs/smb/client/truenas_streams.h:
```diff
@@ -0,0 +1,103 @@
+ * A named stream is a place within a file in addition to the main stream
+ * (normal file data) where data is stored. Named streams have different
+ * data than the main stream (and than each other) and may be written and
+ * read independenly of each other. Named streams for a file are designated
```
Typo: independently

bmeagherix commented on fs/smb/client/truenas_streams.h:
```diff
@@ -0,0 +1,103 @@
+ * ported to FreeBSD and Linux the extattr and xattr OS APIs were layered
+ * on top of the ZFS file-backed xattrs. As time progressed and ZFS on
+ * Linux saw more use, it was determined that the performance and lack of
+ * atomocity of operations on file-backed xattrs was insufficient for
```
Typo: atomicity

bmeagherix commented on fs/smb/client/truenas_streams.h:
```diff
@@ -0,0 +1,103 @@
+ * -----------------
+ * Samba has the ability to present extended attributes as named streams
+ * to SMB clients. This is achieved by prepending a special prefix to
+ * the exended attribute (to differentiate the streams xattrs from normal
```
Typo: extended

bmeagherix commented on fs/smb/client/truenas_streams.h:
```diff
@@ -0,0 +1,103 @@
+ * This commit adds support for the Linux kernel SMB2/3 client to enumerate
+ * streams on a remote SMB server by including them in the output of
+ * listxattr with the special Samba prefix. Streams may be written to
+ * the remote SMB server via setxattr and read through getxtattr. The
```
Typo: getxattr

bmeagherix commented on fs/smb/client/truenas_streams.h:
```diff
@@ -0,0 +1,103 @@
+#define STREAM_XATTR "DosStream."
+#define STREAM_XATTR_LEN (sizeof(STREAM_XATTR) - 1)
+
+#define STREAM_XATTR_PREFIX "user." STREAM_XATTR
+#define STREAM_XATTR_PREFIXLEN (sizeof(STREAM_XATTR_PREFIX) - 1)
```
Maybe add a comment noting that `STREAM_XATTR_PREFIX` is equivalent to samba's `SAMBA_XATTR_DOSSTREAM_PREFIX` - used by `vfs_streams_xattr` module.

anodos325 replied:
Okay. It's actually defined in source3/include/smb.h, so even more generic.

bmeagherix commented on fs/smb/client/truenas_streams.c:
```diff
@@ -0,0 +1,633 @@
+	return rc;
+}
+
+static int
+get_stream_name(const char *name_in, char **name_out)
```
_Maybe_ add a comment saying what we expect inputs and outputs to be,  e.g. "DosStream.streamname:$DATA" -> "streamname"

bmeagherix commented on fs/smb/client/truenas_streams.c:
```diff
@@ -0,0 +1,633 @@
+		return -EINVAL;
+	}
+
+	// chop off :DATA suffix before sending the open request
```
Does this comment need an update?  I see no open request here.

(Then again, you _could_ reference `do_stream_open` and `cifs_convert_stream_path_to_utf16` if you wanted, but meh)

anodos325 replied:
Yeah, this went through a few rounds of refactoring without updating the comment.

bmeagherix commented on fs/smb/client/truenas_streams.c:
```diff
@@ -0,0 +1,633 @@
+	// smaller than this.
+	wsize = tcon->ses->server->ops->wp_retry_size(d_inode(dentry));
+
+	// vfs_streams_xattr in Samba appends an extra NULL byte to stream
+	if (streams_samba_compat_enabled) {
+		sz -= streams_samba_compat_enabled;
+	}
```
This bit of logic isn't that clear to me.  Could you elaborate on the comment please?

Are you saying that a samba server sitting **above** this client will write a trailing NULL that we want to strip off before sending over the wire?  Will anyone _other_ than samba ever write to this?

Besides, even if the code is correct IMO it'd improve readability to simply say:
```
	if (streams_samba_compat_enabled) {
		sz -= 1; /* Remove NULL that was appended by samba */
	}
```
(with whatever tweak to the comment necessary)

anodos325 replied:
That's a good point. I'll redirect user back to notes in the header file and make this clearer.

bmeagherix reviewed (approved):
LGTM, with a few observations:
- Suggested some changes, but they're mostly comment related.
- In terms of the functionality, seems like none of this is actually TrueNAS specific.  So, if we removed the TN rigging, do you think upstream would accept it?  Might be worth inquiring.  (Could point them at this first, and see what is said.  It's a fairly corner-case piece of functionality, so they simply may _not_ be interested - e.g. maintenance burden)

amotin commented on fs/smb/client/truenas_streams.c:
```diff
@@ -0,0 +1,633 @@
+		 */
+		name = (char *)stream + sizeof(struct smb2_file_stream_info);
+
+		stream_name = cifs_strndup_from_utf16(name, stream->StreamNameLength, true,
+						      cifs_sb->local_nls);
```
Would it make sense to check here or in get_next_stream() that `name + stream->StreamNameLength <= end_of_streams`?

amotin commented on fs/smb/client/truenas_streams.c:
```diff
@@ -0,0 +1,633 @@
+			break;
+	}
+
+	if (streams_samba_compat_enabled) {
+		// Append terminating NULL to match Samba behavior
+		*(dst + total_read + 1) = '\0';
+	}
+
+	SMB2_close(xid, tcon, fid.persistent_fid, fid.volatile_fid);
+	if (rc == 0) {
+		if (streams_samba_compat_enabled) {
+			*pused = total_read + streams_samba_compat_enabled;
+			*(dst + total_read) = '\0';
```
Don't you set the NULL twice and for some reason even at different offsets?  Should the first `if` be removed?

amotin commented on fs/smb/client/truenas_streams.c:
```diff
@@ -0,0 +1,633 @@
+		return rc;
+	}
+
+	if ((value == NULL) && (size == 0)) {
```
I am not sure what is the semantics here, but checking both and with `&&` looks redundant and potentially unsafe here.  I would check one, and if needed, asserted another.

amotin commented on fs/smb/client/smb2pdu.c:
```diff
@@ -3772,6 +3772,20 @@ int SMB2_query_info(const unsigned int xid, struct cifs_tcon *tcon,
+	u32 *plen)
+{
+	*plen = 0;
+	return query_info(xid, tcon, persistent_fid, volatile_fid,
+			  FILE_STREAM_INFORMATION, SMB2_O_INFO_FILE, 0,
+			  (sizeof(struct smb2_file_stream_info) + 255) * 32,
+			  sizeof(struct smb2_file_stream_info), (void **)data,
+			  plen);
```
> Stream names may be no more than 255 characters in length

Is that 255 UTF-16 "characters" or bytes? SMB2_query_streams() uses some 255 constant, which seems like in bytes. Other code parts in SMB clients I see multiply the length by 2 before writing to StreamNameLength, so odd 255 value there looks odd. :slightly_smiling_face:  Also I see your own comment:

 * NOTE regarding aligment:
 * When multiple FILE_STREAM_INFORMATION data elements are present in the
 * buffer, each MUST be alinged on an 8-byte boundaries; any bytes inserted
 * for alignment SHOULD be set to zero and the receiver MUST ignore them.

Does it mean it should at least be 256 there, or may be even 512 or more?

anodos325 committed:
Address review

anodos325 committed:
Various comment fixes

anodos325 committed:
Address review

anodos325 committed:
Enable large xattr support for SMB client mounts

anodos325 committed:
Transition many server operations to be protocol-agnostic

This partially provides what we will need to make ADS code
protocol-agnostic and potentially allow SMB1 support. Further
work is required to handle opens.

anodos325 committed:
Allow bypassing MAX_EA_SIZE restriction for streams

anodos325 committed:
Fix listxattr for case where no streams are present

anodos325 committed:
Allow SMB 2.10 protocol to interact with streams

bugclerk commented:
Not updating JIRA ticket https://ixsystems.atlassian.net/browse/NAS-127114 target versions as no JIRA version corresponds to this PR

bugclerk commented:
This PR has been merged and conversations have been locked.
If you would like to discuss more about this issue please use our forums or raise a Jira ticket.
